### PR TITLE
Implemented 'open-recent'

### PR DESCRIPTION
**Pull request by shalabhc**
Shows a list of recently closed files. The list persists across restarts.

### CODE REVIEW

**Comment by shalabhc** [comment]
@nilnor - correct. Limiting the list to some size is probably a good idea. What do you mean by pruning here - manual deletion of history or automatic?

**Comment by nilnor** [comment]
Sorry, I was unclear. I believe I was thinking along the lines that it could either a hard limit, or it could be limited by how long ago the file was opened (i.e. prune entries older than X). Anyhow, I think a simple size limit will be the best.

**Comment by shalabhc** [comment]
OK, added size limit - ready for review!

File: lib/howl/application.moon [hunk 1]
```diff
@@ -4,13 +4,20 @@
...
 import Window, Editor, theme from howl.ui
-import Buffer, Settings, mode, bundle, bindings, keymap, signal, interact, timer, clipboard from howl
+import Buffer, Settings, mode, bundle, bindings, keymap, signal, interact, timer, clipboard, config from howl
 import File, Process from howl.io
 import PropertyObject from howl.aux.moon
 Gtk = require 'ljglibs.gtk'
 
 append = table.insert
 
+config.define
```
Comment (nilnor) [inline lib/howl/application.moon]: As a style nit-picking I'd prefer if this was moved down at the end of the file instead, together with the signal definitions.

File: lib/howl/application.moon [hunk 2]
```diff
@@ -116,6 +126,18 @@ class Application extends PropertyObject
...
     @_buffers = [b for b in *@_buffers when b != buffer]
 
+    if buffer.file
+      @_recently_closed = [file_info for file_info in *@_recently_closed when file_info.file != buffer.file]
+      append @_recently_closed, {
+        file: buffer.file
+        last_shown: buffer.last_shown
+      }
+      count = #@_recently_closed
+      limit = howl.config.recently_closed_limit
```
Comment (nilnor) [inline lib/howl/application.moon]: `config` is already in scope here, so need for `howl.config`.

File: lib/howl/application.moon [hunk 3]
```diff
@@ -116,6 +126,18 @@ class Application extends PropertyObject
 
     @_buffers = [b for b in *@_buffers when b != buffer]
 
+    if buffer.file
+      @_recently_closed = [file_info for file_info in *@_recently_closed when file_info.file != buffer.file]
```
Comment (nilnor) [inline lib/howl/application.moon]: This block could be its own private method instead, as to not clutter `close_buffer` to much.

File: lib/howl/application.moon [hunk 4]
```diff
@@ -311,7 +341,7 @@ class Application extends PropertyObject
   _restore_session: (window, restore_buffers) =>
     session = @settings\load_system 'session'
 
-    if session and session.version == 1
+    if session and session.version >= 1
```
Comment (nilnor) [inline lib/howl/application.moon]: I wonder whether we should skip the version check completely for now, as we don't have any breaking incompatible changes yet. We could just check for the existence of `session.recently_closed` below. As it is, the version check seems wrong as it now accepts any future version of the session.

File: lib/howl/keymap.moon [hunk 5]
```diff
@@ -77,6 +77,7 @@
   }
 
   ctrl_o:           'open'
+  ctrl_shift_o:     'open-recent'
```
Comment (nilnor) [inline lib/howl/keymap.moon]: We should have a binding for OS X as well I guess?

File: lib/howl/commands/file_commands.moon [hunk 6]
```diff
@@ -39,6 +39,25 @@ command.register
...
 
+command.register
+  name: 'open-recent',
+  description: 'Open a recently closed file'
+  input: ->
+    recent_files = {}
+    for file_info in *app.recently_closed
+      table.insert recent_files, {
+        file_info.file.basename
+        file_info.file.parent.path
```
Comment (nilnor) [inline lib/howl/commands/file_commands.moon]: Use `.short_path` here for contracting the home dir.

**Comment by nilnor** [comment]
All done!

I'll address your idea of changing the semantics to just keep track of all the recently opened files brought up in #134 here - I think it's a good idea. It makes the implementation simpler as well, which is good. The only question I can think of there is what the command should be named, as it's a combo of switch and open.

**Comment by shalabhc** [comment]
I'm not sure it makes the implementation simpler. I would just leave `app.recently_closed` as is but use both `app.buffers` and `app.recently_closed` in the command itself.

As for naming - I don't see a problem with `open-recent` as one could think of it as opening a file rather than a buffer. This is similar to selecting an already open file in the `open` command.

**Comment by shalabhc** [comment]
All done - ready for review again!

**Comment by nilnor** [comment]
Merged!

File: lib/howl/commands/file_commands.moon [hunk 7]
```diff
@@ -39,6 +39,31 @@ command.register
   input: -> interact.select_file allow_new: true
   handler: (file) -> app\open_file file
 
+command.register
```
Comment (shalabhc) [inline lib/howl/commands/file_commands.moon]: @nilnor - this branch was merged but the latest change to the command didn't get in? Specifically the buffers are now displayed and the title is not updated.

Comment (shalabhc) [inline lib/howl/commands/file_commands.moon]: Hmm, it's not just this - maybe the last few commits didn't make it in, including the OSX binding and implementation cleanup.

**Pull request closed without merging**